### DISCUSSION TITLE
Podman ignores FIPS policy modifiers

### DISCUSSION THREAD

**Discussion by ihatethecloud** [containers/podman#23213] [General]
https://github.com/containers/podman/blob/b5bfd7233b1a3d500f0df1968dfcc2bd5c3c30ce/vendor/github.com/containers/common/pkg/subscriptions/subscriptions.go#L347-L395

podman-4.9.4-1 on almalinux 8.10

If the user has a crypto-policy modifier there is no way for him to have it inside the container.

1 - Create a policy modifier 
```
/usr/share/crypto-policies/policies/modules/TEST-MOD.pmod
cipher = -AES-256-CBC
```

2 - Activate the policy
```
update-crypto-policies --set FIPS:TEST-MOD
```

3 - Reboot
4 - Check crypto-policy
```
cat /etc/crypto-policies/config
FIPS:TEST-MOD
```
5 - Check inside container
```
podman run --rm --it almalinux:8.9 cat /etc/crypto-policies/config
FIPS
```
6 - Try with bind volume
```
podman run --rm -v /etc/crypto-policies/:/etc/crypto-policies/ --it almalinux:8.9 cat /etc/crypto-policies/config
FIPS
```

**Comment by rhatdan** [comment]
$ cat /tmp/crypto-policies/config 
FIPS:TEST-MOD
 $ podman run --rm -v /tmp/crypto-policies:/etc/crypto-policies:Z -it almalinux:8.9 cat /etc/crypto-policies/config
FIPS
$ podman run --rm -v /tmp/crypto-policies/config:/etc/crypto-policies/config:Z -it almalinux:8.9 cat /etc/crypto-policies/config
FIPS:TEST-MOD

You would like to change the check in [podman/vendor/github.com/containers/common/pkg/subscriptions/subscriptions.go](https://github.com/containers/podman/blob/b5bfd7233b1a3d500f0df1968dfcc2bd5c3c30ce/vendor/github.com/containers/common/pkg/subscriptions/subscriptions.go#L347-L395)

To also check if /etc/crypto-policies is a mount point?

**Reply by ihatethecloud** [reply]
It would be ideal if podman gives way to bind volumes instead of enforcing the crypto-policy without modifiers.

**Reply by neverpanic** [reply]
This is working as intended, and what you are asking for is a feature request that can probably not be fulfilled.

The crypto-policy active on the host was never intended to be copied to the container. `podman` cannot know whether the container to be started uses the same or a compatible set of crypto-policies (you could be running a RHEL 8 host but an AlmaLinux 9 container). The files in `/etc/crypto-policies/back-ends` are specific to the versions of the cryptographic libraries used on a machine, so bind-mounting them into a container could either lead to unexpected side-effects, unexpected behavior, or in the worst case, a system that doesn't work because a major cryptographic library cannot parse its configuration file.

Even if `podman` knew that a `TEST-MOD` policy file existed in the container to be started, the only safe way to interpret that policy file would be to invoke the `update-crypto-policies` utility *inside* of that container to parse it and generate the correct configuration.

If you want to apply a subpolicy to the `FIPS` policy in a container, the only way to do that that I can see is to build a container image that has this policy set already, and then disable `podman`'s automatic bind-mounts. `MountsWithUIDGID` at https://github.com/containers/podman/blob/main/vendor/github.com/containers/common/pkg/subscriptions/subscriptions.go#L189 contains a `disableFips` parameter that could be used for this, but IIRC the last time I checked, this parameter was only modifiable in OpenShift, not in podman (https://github.com/containers/podman/blob/main/libpod/container_internal_common.go#L2054)

**Reply by neverpanic** [reply]
@rhatdan Also note that the test you did above:
```
$ podman run --rm -v /tmp/crypto-policies/config:/etc/crypto-policies/config:Z -it almalinux:8.9 cat /etc/crypto-policies/config
FIPS:TEST-MOD
```
Does actually *NOT* put the system into the crypto-policy `FIPS:TEST-MOD`. You overwrite the internal file that `update-crypto-policies --show` would have used to tell you what policy it is in, but the actual configuration files for the cryptographic libraries in `/etc/crypto-policies/back-ends` would still have been those generated by the `FIPS` crypto-policy.

**Reply by t184256** [reply]
+1 to what @neverpanic says. I can't say I'm 100% sure what were you expecting to happen, but you expected container to automatically follow host policy, be it in the name or the effective policy contents, this is simultaneously undesireable and impossible without running policy generation during the container spin-up or something, and that's if the policy exists / all the keywords are supported by the container version of crypto-policies.

**Comment by ihatethecloud** [comment]
As a work around I recompiled podman without the code which enforces fips on the container. Mounting /etc/crypto-policies with bind volumes works really good.

@rhatdan: Can we expect a solution to the discussion in matter?

**Comment by rhatdan** [comment]
Interested in opening a PR?

**Comment by robbmanes** [comment]
Looks like this might be resolved as per https://github.com/containers/common/commit/abf2f6bc69b86139f5915a6723c4cb35d241ef21#diff-5a9cb9e0ff686c88747667916a4873403ef88b4947af001d5edf24841157106cR358-R362
